src/main/scala/uk/gov/hmrc/smartstub/AutoGen.scala

```diff
@@ -87,53 +127,76 @@ object AutoGen extends LowPriorityGenProviderInstances {
     _ => instance(Gen.oneOf(true,false))
 
   // generic instance
+  implicit def providerGeneric[A]
+  (implicit
+   generic: Mirror.ProductOf[A],
+   hGenProvider: => GenProvider[generic.MirroredElemTypes]
+  ): GenProvider[A] =
+    instance(hGenProvider.gen.map(t => generic.fromProduct(t)))
+  extension [A](mirror: Mirror.SumOf[A])
+    def safeFromOrdinal(i: Int): A =
+      mirror.asInstanceOf[{ def fromOrdinal(i: Int): A}].fromOrdinal(i)
 
-  implicit def providerGeneric[A, H, T]
+  implicit def providerSum[A]
   (implicit
-   generic: LabelledGeneric.Aux[A,T],
-   hGenProvider: Lazy[GenProvider[T]]
+   s: Mirror.SumOf[A],
+   sumProvider: SumGenProvider[s.MirroredElemTypes]
   ): GenProvider[A] =
-    instance(hGenProvider.value.gen.map(generic.from))
+    GenProvider.instance(
+    sumProvider.gen.map{ i => s.safeFromOrdinal(i)}
+  )
 
   // HList instances
+  type FieldType[K, V] = V
 
-  implicit val providerHNil: GenProvider[HNil] = instance(Gen.const(HNil))
+  implicit def genProviderByFieldName[H](implicit gp: GenProvider[H]): String => GenProvider[H] =
+    (_: String) => gp
 
-  implicit def providerHCons[K <: Symbol, H, T <: HList]
+  implicit val providerHNil: GenProvider[EmptyTuple] = instance(Gen.const(EmptyTuple))
+
+  implicit def providerHCons[K <: String & Singleton, H, T <: Tuple]
   (implicit
-   witness: Witness.Aux[K],
-   hGenProvider: Lazy[String => GenProvider[H]],
-   tGenProvider: Lazy[GenProvider[T]]
-  ): GenProvider[FieldType[K,H] :: T] = instance(
-    hGenProvider.value(witness.value.name).gen.flatMap(f =>
-      tGenProvider.value.gen.map{ t =>
-        field[K](f) :: t
+   valueOfK: ValueOf[K],
+   hGenProvider: String => GenProvider[H],
+   tGenProvider: => GenProvider[T]
+  ): GenProvider[H *: T] = {
+    val fieldName: String = valueOfK.value
+    val headGenerator = hGenProvider(fieldName).gen.flatMap(f =>
+      tGenProvider.gen.map { t =>
+        f *: t
       }
     )
-  )
+    instance(headGenerator)
+  }
 
```

Comment:
Both providerCCons methods. have duplicated logic. we can change it as 
<img width="939" alt="image" src="https://github.com/user-attachments/assets/93c06b90-34c6-4df1-8351-f95b542bd380" />
<img width="951" alt="image" src="https://github.com/user-attachments/assets/e069f3c5-0ca8-4d6c-9747-a04854e95504" />
check if able to refactor as this?